README.md
# defexception
[![Build Status](https://travis-ci.org/redplanetlabs/defexception.svg?branch=master)](https://travis-ci.org/redplanetlabs/defexception)
[![Clojars Project](https://img.shields.io/clojars/v/com.rpl/defexception.svg)](https://clojars.org/com.rpl/defexception)
[![cljdoc
badge](https://cljdoc.org/badge/com.rpl/defexception)](https://cljdoc.org/d/com.rpl/defexception/CURRENT)

A simple Clojure library that allows one to dynamically define Java Exception classes in Clojure.

## Motivation

Sometimes you just want a quick Exception class that you can catch
directly. `defexception` uses Java bytecode to dynamically create a
Java class that directly inherits from `clojure.lang.ExceptionInfo`.

The created exception class does not add any behavior to `clojure.lang.ExceptionInfo`.

## Dependency


[deps.edn](https://clojure.org/guides/deps_and_cli) dependency information:

```
com.rpl/defexception {:mvn/version "0.2.5"}
```

[Leiningen](https://github.com/technomancy/leiningen/) dependency information:

```
[com.rpl/defexception "0.2.5"]
```

## Usage

To create your own exception class you can do this:

```clojure
(ns foo.bar
 (:require [com.rpl.defexception :refer [defexception]]))

(defexception MyException)
```

This will create the `foo.bar.MyException` class that inherits from
`clojure.lang.ExceptionInfo`. This will also `import` the class into
the current namespace and create a helper function
`foo.bar/->MyException` to help you construct the exception class.

Now you can do this:

```clojure
(try
  (throw (->MyException "My bad!" {:my-bad 1}))
  (catch MyException e
    (ex-data e)))
;; => {:my-bad 1}
```

The generated `->MyException` helper function has several signatures
to help you instantiate your exception.

```clojure
;; creates an exception with no message or ex-data
(->MyException)

;; creates an exception with only ex-data
(->MyException {:hello 1})

;; creates an exception with only a message
(->MyException "My Bad!")

;; creates an exception with both a message and ex-data
(->MyException "My Bad!" {:hello 1})

;; creates an exception with a message, ex-data and a cause
(->MyException "My Bad!" {:hello 1} (Exception. "The cause"))
```

## AOT compilation

Care was taken to make these exceptions compatible with Clojure's AOT
compilation.

## License

Copyright © 2019 - 2020 Red Planet Labs Inc.

Distributed under the Eclipse Public License version 1.0


CHANGELOG.md
# Changelog

## [Unreleased]

## [0.2.5] — 2022-11-04
### Fixed
- Use boolean, not instance, in constructor cache key

## [0.2.4] — 2021-09-28
### Fixed
- release using jdk 1.8

## [0.2.3] — 2021-09-28
### Fixed
- removed reflection

## [0.2.2] — 2020-12-23
### Changed
- Add IDefException interface on generated exception classes

## [0.2.1] — 2020-08-26
### Changed
- update to clojure 1.10.1

## [0.2.0] - 2020-01-29
### Changed
- Make the stack trace start at the point of instantiation
- Override toString to include the class name

## 0.1.0 - 2019-08-22

[0.2.0]: https://github.com/redplanetlabs/defexception/compare/0.1.0...0.2.0
[0.2.1]: https://github.com/redplanetlabs/defexception/compare/0.2.0...0.2.1
[0.2.2]: https://github.com/redplanetlabs/defexception/compare/0.2.1...0.2.2
[0.2.3]: https://github.com/redplanetlabs/defexception/compare/0.2.2...0.2.3
[0.2.4]: https://github.com/redplanetlabs/defexception/compare/0.2.3...0.2.4
[0.2.5]: https://github.com/redplanetlabs/defexception/compare/0.2.4...0.2.5
[Unreleased]: https://github.com/redplanetlabs/defexception/compare/0.2.5...HEAD


test/defexception/core_test.clj
(ns defexception.core-test
  (:require [clojure.test :refer :all]
            [defexception.core :refer [defexception]]
            [clojure.java.io :as io]))

(defexception MyException)

(binding [*compile-files* true]
  (defexception TestException))

(deftest dynamic-exception-class-test
  (let [ex (MyException. "Message" {:hello 1})]
    (is (= "Message" (.getMessage ex)))
    (is (= {:hello 1} (ex-data ex))))
  (let [cause (Exception. "Cause")
        ex (MyException. "Message" {:hello 1} cause)]
    (is (= "Message" (.getMessage ex)))
    (is (= {:hello 1} (ex-data ex)))
    (is (= cause (.getCause ex)))))

(deftest dynamic-exception-class-const-fn
  (let [ex (->MyException)]
    (is (= nil (.getMessage ex)))
    (is (= {} (ex-data ex))))
  (let [ex (->MyException "Message")]
    (is (= "Message" (.getMessage ex)))
    (is (= {} (ex-data ex))))
  (let [ex (->MyException {:hello 1})]
    (is (= nil (.getMessage ex)))
    (is (= {:hello 1} (ex-data ex))))
  (let [ex (->MyException "Message" {:hello 1})]
    (is (= "Message" (.getMessage ex)))
    (is (= {:hello 1} (ex-data ex))))
  (let [cause (Exception. "Cause")
        ex (->MyException "Message" {:hello 1} cause)]
    (is (= "Message" (.getMessage ex)))
    (is (= {:hello 1} (ex-data ex)))
    (is (= cause (.getCause ex)))))

(deftest has-aoted
  (is (..
       (io/file *compile-path* "defexception/core_test/TestException.class")
       exists)))

(deftest aot-class-test
  (let [ex (->TestException)]
    (is (= nil (.getMessage ex)))
    (is (= {} (ex-data ex))))
  (let [ex (->TestException "Message")]
    (is (= "Message" (.getMessage ex)))
    (is (= {} (ex-data ex))))
  (let [ex (->TestException {:hello 1})]
    (is (= nil (.getMessage ex)))
    (is (= {:hello 1} (ex-data ex))))
  (let [ex (->TestException "Message" {:hello 1})]
    (is (= "Message" (.getMessage ex)))
    (is (= {:hello 1} (ex-data ex))))
  (let [cause (Exception. "Cause")
        ex (->TestException "Message" {:hello 1} cause)]
    (is (= "Message" (.getMessage ex)))
    (is (= {:hello 1} (ex-data ex)))
    (is (= cause (.getCause ex)))))


test/defexception/other_test.clj
(ns defexception.other-test
  (:require [clojure.test :refer :all]
            [defexception.core-test :refer [->TestException]])
  (:import [defexception.core_test TestException]))

;; this is simply testing the import of an exception from a namespace
;; where it's defined

(deftest aot-class-test
  (let [ex (->TestException)]
    (is (= nil (.getMessage ex)))
    (is (= {} (ex-data ex))))
  (let [ex (->TestException "Message")]
    (is (= "Message" (.getMessage ex)))
    (is (= {} (ex-data ex))))
  (let [ex (->TestException {:hello 1})]
    (is (= nil (.getMessage ex)))
    (is (= {:hello 1} (ex-data ex))))
  (let [ex (->TestException "Message" {:hello 1})]
    (is (= "Message" (.getMessage ex)))
    (is (= {:hello 1} (ex-data ex))))
  (let [cause (Exception. "Cause")
        ex (->TestException "Message" {:hello 1} cause)]
    (is (= "Message" (.getMessage ex)))
    (is (= {:hello 1} (ex-data ex)))
    (is (= cause (.getCause ex)))))





test/com/rpl/defexception_test.clj
(ns com.rpl.defexception-test
  (:require [clojure.test :refer :all]
            [com.rpl.defexception :refer [defexception]]
            [clojure.java.io :as io])
  (:import [com.rpl.defexception IDefException]))

(defexception MyException)

(binding [*compile-files* true]
  (defexception TestException))

(deftest dynamic-exception-class-test
  (let [ex (MyException. "Message" {:hello 1})]
    (is (instance? IDefException ex))
    (is (= "Message" (.getMessage ex)))
    (is (= {:hello 1} (ex-data ex)))
    (is (= "MyException: Message {:hello 1}" (str ex))))
  (let [cause (Exception. "Cause")
        ex (MyException. "Message" {:hello 1} cause)]
    (is (not (instance? IDefException cause)))
    (is (= "Message" (.getMessage ex)))
    (is (= {:hello 1} (ex-data ex)))
    (is (= cause (.getCause ex)))
    (is (= "MyException: Message {:hello 1}" (str ex)))))

(deftest dynamic-exception-class-const-fn
  (let [ex (->MyException)]
    (is (= nil (.getMessage ex)))
    (is (= {} (ex-data ex)))
    (is (= "MyException: null {}" (str ex))))
  (let [ex (->MyException "Message")]
    (is (= "Message" (.getMessage ex)))
    (is (= {} (ex-data ex)))
    (is (= "MyException: Message {}" (str ex))))
  (let [ex (->MyException {:hello 1})]
    (is (= nil (.getMessage ex)))
    (is (= {:hello 1} (ex-data ex)))
    (is (= "MyException: null {:hello 1}" (str ex))))
  (let [ex (->MyException "Message" {:hello 1})]
    (is (= "Message" (.getMessage ex)))
    (is (= {:hello 1} (ex-data ex)))
    (is (= "MyException: Message {:hello 1}" (str ex))))
  (let [cause (Exception. "Cause")
        ex (->MyException "Message" {:hello 1} cause)]
    (is (= "Message" (.getMessage ex)))
    (is (= {:hello 1} (ex-data ex)))
    (is (= cause (.getCause ex)))
    (is (= "MyException: Message {:hello 1}" (str ex)))))

(deftest has-aoted
  (is (..
       (io/file *compile-path* "com/rpl/defexception_test/TestException.class")
       exists)))

(deftest aot-class-test
  (let [ex (->TestException)]
    (is (= nil (.getMessage ex)))
    (is (= {} (ex-data ex)))
    (is (= "TestException: null {}" (str ex))))
  (let [ex (->TestException "Message")]
    (is (= "Message" (.getMessage ex)))
    (is (= {} (ex-data ex)))
    (is (= "TestException: Message {}" (str ex))))
  (let [ex (->TestException {:hello 1})]
    (is (= nil (.getMessage ex)))
    (is (= {:hello 1} (ex-data ex)))
    (is (= "TestException: null {:hello 1}" (str ex))))
  (let [ex (->TestException "Message" {:hello 1})]
    (is (= "Message" (.getMessage ex)))
    (is (= {:hello 1} (ex-data ex)))
    (is (= "TestException: Message {:hello 1}" (str ex))))
  (let [cause (Exception. "Cause")
        ex (->TestException "Message" {:hello 1} cause)]
    (is (= "Message" (.getMessage ex)))
    (is (= {:hello 1} (ex-data ex)))
    (is (= cause (.getCause ex)))
    (is (= "TestException: Message {:hello 1}" (str ex)))))

(defmacro line [] (-> &form meta :line))
(defmacro cur-file [] *file*)

(deftest stack-trace-test
  (let [ex (->TestException)
        instantiation-line-no (dec (line))]
    (try
      (throw ex)
      (catch TestException e
        (let [first-stack-trace-element (first (.getStackTrace e))
              line-no (.getLineNumber first-stack-trace-element)
              file-name (.getFileName first-stack-trace-element)]
          (is (= instantiation-line-no line-no))
          (is (= (last (clojure.string/split (cur-file) #"/")) file-name))
          ))))
  )


src/defexception/core.clj
(ns defexception.core
  (:require [clojure.string :as string]
            [defexception.impl]))

(defmacro defexception
  "Dynamically creates an clojure.lang.ExceptionInfo class using JVM
  bytecode. The exception class inherits its behavior from
  clojure.lang.ExceptionInfo yet will be its own type.

  This means that an instance of this exception will respond to
  `clojure.core/ex-data` and can have an optional Throwable `cause` arg.

  This will create a constructor function much similar to the one
  created for Records.

  Example Usage:

  (defexception MyException) => user.MyException

  (->MyException)
  (->MyException {:hello 1})
  (->MyException \"This is a message\")
  (->MyException \"This is a message\" {:hello 1})
  (->MyException \"This is a message\" {:hello 1} (Exception. \"A cause\"))

  (ex-data (->MyException {:hello 1})) => {:hello 1}

  (ex-data (->MyException \"This is a message\" {:hello 1})) => {:hello 1}

  (.getMessage (->MyException \"This is a message\")) => \"This is a message\"

  or invoke the constructors directly

  (user/MyException. \"This is a Message\" {})
  (user/MyException. \"This is a Message\" {} (Exception. \"A message\"))"
  [t]
  (let [class-name (str (string/replace (str *ns*) "-" "_") "." t)]
    `(let [x# (defexception.impl/load-or-mk-ex-info-class ~class-name)]
       (import ~(symbol class-name))
       (defn ~(symbol (str "->" t))
         ([] (defexception.impl/make-ex x# nil {} nil))
         ([~'msg-o-data]
          (if (map? ~'msg-o-data)
            (defexception.impl/make-ex x# nil ~'msg-o-data nil)
            (defexception.impl/make-ex x# ~'msg-o-data {} nil)))
         ([~'msg ~'data] (defexception.impl/make-ex x# ~'msg ~'data nil))
         ([~'msg ~'data ~'cause] (defexception.impl/make-ex x# ~'msg ~'data ~'cause)))
       x#)))


src/com/rpl/defexception.clj
(ns com.rpl.defexception
  (:require [clojure.string :as string]
            [com.rpl.defexception.impl :as impl]))

(defmacro defexception
  "Dynamically creates an clojure.lang.ExceptionInfo class using JVM
  bytecode. The exception class inherits its behavior from
  clojure.lang.ExceptionInfo yet will be its own type.

  This means that an instance of this exception will respond to
  `clojure.core/ex-data` and can have an optional Throwable `cause` arg.

  This will create a constructor function much similar to the one
  created for Records.

  Example Usage:

  (defexception MyException) => user.MyException

  (->MyException)
  (->MyException {:hello 1})
  (->MyException \"This is a message\")
  (->MyException \"This is a message\" {:hello 1})
  (->MyException \"This is a message\" {:hello 1} (Exception. \"A cause\"))

  (ex-data (->MyException {:hello 1})) => {:hello 1}

  (ex-data (->MyException \"This is a message\" {:hello 1})) => {:hello 1}

  (.getMessage (->MyException \"This is a message\")) => \"This is a message\"

  or invoke the constructors directly

  (user/MyException. \"This is a Message\" {})
  (user/MyException. \"This is a Message\" {} (Exception. \"A message\"))"
  [t]
  (let [class-name (str (string/replace (str *ns*) "-" "_") "." t)]
    `(let [x# (impl/load-or-mk-ex-info-class ~class-name)]
       (import ~(symbol class-name))
       (defn ~(symbol (str "->" t))
         ([] (impl/make-ex x# ~(name t) {} nil))
         ([~'msg-o-data]
          (if (map? ~'msg-o-data)
            (impl/make-ex x# ~(name t) ~'msg-o-data nil)
            (impl/make-ex x# ~'msg-o-data {} nil)))
         ([~'msg ~'data] (impl/make-ex x# ~'msg ~'data nil))
         ([~'msg ~'data ~'cause] (impl/make-ex x# ~'msg ~'data ~'cause)))
       x#)))


src/clj/com/rpl/defexception/impl.clj
(ns ^:no-doc com.rpl.defexception.impl
  (:require [clojure.string :as string])
  (:import [clojure.asm MethodVisitor
            ClassVisitor ClassWriter Opcodes Type]
           [clojure.asm.commons GeneratorAdapter Method]
           [clojure.lang ExceptionInfo]
           [com.rpl.defexception IDefException]))

(defn- forward-constructor [^ClassWriter cw ^Type t ^Method constr]
  (doto (GeneratorAdapter. Opcodes/ACC_PUBLIC constr nil nil cw)
    (.loadThis)
    (.loadArgs)
    (.invokeConstructor t constr)
    (.returnValue)
    (.endMethod)))

(defn- to-string-method [^ClassWriter cw ^String exception-class-name]
  (let [string (Type/getType String)
        string-builder (Type/getType StringBuilder)
        string-append (Method/getMethod
                        "java.lang.StringBuilder append (java.lang.String)")
        to-string (Method/getMethod "java.lang.String toString ()")]
    (doto (GeneratorAdapter. Opcodes/ACC_PUBLIC to-string nil nil cw)
      (.newInstance string)
      (.dup)

      (.newInstance string-builder)
      (.dup)
      (.invokeConstructor string-builder (Method/getMethod "void <init> ()"))

      (.push (str exception-class-name ": "))
      (.invokeVirtual string-builder string-append)

      (.loadThis)
      (.invokeVirtual (Type/getType Exception) (Method/getMethod "String getMessage ()"))
      (.invokeVirtual string-builder string-append)

      (.push " ")
      (.invokeVirtual string-builder string-append)

      (.loadThis)
      (.getField (Type/getType clojure.lang.ExceptionInfo)
                 "data"
                 (Type/getType clojure.lang.IPersistentMap))
      (.invokeVirtual (Type/getType Object) to-string)
      (.invokeVirtual string-builder string-append)

      (.invokeConstructor
        string
        (Method/getMethod "void <init> (java.lang.StringBuilder)"))

      (.returnValue)
      (.endMethod))))

(defn- define-class [^clojure.lang.DynamicClassLoader cl ^String name ^ClassWriter cw]
  (let [klass (.defineClass cl name (.toByteArray cw) nil)]
    (when *compile-files*
      (clojure.lang.Compiler/writeClassFile
       (string/replace (munge (.getName klass)) "." "/")
       (.toByteArray cw)))
    klass))

(defn- mk-ex-info-class [exception-class-name]
  #_(prn :making exception-class-name)
  (let [cw (ClassWriter. (+ ClassWriter/COMPUTE_MAXS ClassWriter/COMPUTE_FRAMES))
        ex-info-type (Type/getType clojure.lang.ExceptionInfo)
        internal-name (string/replace exception-class-name "." "/")
        class-name (-> internal-name (string/split #"/") last)]
    (.visit cw
            ;; this allows compatibility back to Clojure 1.4
            ;; and is = clojure.asm.Opcodes/V1_7
            (inc clojure.asm.Opcodes/V1_6)
            (+ Opcodes/ACC_SUPER Opcodes/ACC_PUBLIC)
            internal-name
            nil
            (.getInternalName ex-info-type)
            (into-array String [(Type/getInternalName IDefException)]))
    (forward-constructor
      cw
      ex-info-type
      (Method/getMethod "void <init> (String, clojure.lang.IPersistentMap)"))
    (forward-constructor
      cw
      ex-info-type
      (Method/getMethod "void <init> (String, clojure.lang.IPersistentMap, Throwable)"))
    (to-string-method
      cw
      class-name)
    (.visitEnd cw)
    (define-class
      (clojure.lang.DynamicClassLoader.)
      exception-class-name
      cw)))

(defn load-or-mk-ex-info-class [exception-class-name]
  (try (clojure.lang.RT/classForName exception-class-name)
       (catch ClassNotFoundException e
         (mk-ex-info-class exception-class-name))))

(def ^:private ex-info-const
  (memoize
   (fn [^Class klass cause?]
     (assert (boolean? cause?))
     (let [arg-types [String clojure.lang.IPersistentMap]]
       (.getConstructor
        klass
        (into-array Class
                    (if cause?
                      (conj arg-types Throwable)
                      arg-types)))))))

(defn- fix-stack-trace [^ExceptionInfo ex-info-instance]
  (let [stacktrace (.getStackTrace ex-info-instance)
        constructor-pattern (re-pattern
                             (str ".*"
                                  (munge
                                   (str "->"
                                        (-> ex-info-instance
                                            class
                                            (.getSimpleName))))
                                  ".*"))
        updated-stacktrace (->> stacktrace
                                (drop-while
                                 (fn [^StackTraceElement elt]
                                   (not
                                    (re-matches
                                     constructor-pattern
                                     (.getClassName elt)))))
                                rest
                                (into-array StackTraceElement))]
    (when stacktrace
      (.setStackTrace ex-info-instance updated-stacktrace))
    ex-info-instance)
  )

(defn make-ex
  "use reflection to instantiate the exception class"
  [^Class klass msg data cause]
  (let [constr ^java.lang.reflect.Constructor (ex-info-const klass (boolean cause))
        args [msg (or data {})]]
    (fix-stack-trace (.newInstance constr
                                   (object-array
                                    (if cause
                                      (conj args cause)
                                      args))))))


project.clj
(defproject com.rpl/defexception "0.2.5"
  :description "A simple Clojure library that allows one to dynamically define Java Exception classes in Clojure."
  :url "https://github.com/redplanetlabs/defexception"
  :license {:name "Eclipse Public License"
            :url  "http://www.eclipse.org/legal/epl-v10.html"}
  :scm {:name "git" :url "https://github.com/redplanetlabs/defexception"}
  :dependencies []
  :source-paths ["src/clj"]
  :java-source-paths ["src/java"]
  :plugins [#_[lein-cloverage "1.0.13"]
            [lein-shell "0.5.0"]
            [lein-ancient "0.6.15"]
            [lein-changelog "0.3.2"]]
  :profiles {:dev {:dependencies [[org.clojure/clojure "1.10.1"]]}
             :1.2 {:dependencies [[org.clojure/clojure "1.2.0"]]}
             :1.3 {:dependencies [[org.clojure/clojure "1.3.0"]]}
             :1.4 {:dependencies [[org.clojure/clojure "1.4.0"]]}
             :1.5 {:dependencies [[org.clojure/clojure "1.5.0"]]}
             :1.6 {:dependencies [[org.clojure/clojure "1.6.0"]]}
             :1.7 {:dependencies [[org.clojure/clojure "1.7.0"]]}
             :1.8 {:dependencies [[org.clojure/clojure "1.8.0"]]}
             :1.9 {:dependencies [[org.clojure/clojure "1.9.0"]]}}
  :deploy-repositories [["releases" :clojars]]
  :aliases {"update-readme-version"
            ["shell" "sed" "-I" ".bak"
             "-e" "s/defexception \"[0-9.]*\"/defexception \"${:version}\"/"
             "-e" "s/version \"[0-9.]*\"/version \"${:version}\"/"
             "README.md"]}
  :release-tasks [["shell" "git" "diff" "--exit-code"]
                  ["change" "version" "leiningen.release/bump-version"]
                  ["change" "version" "leiningen.release/bump-version" "release"]
                  ["changelog" "release"]
                  ["update-readme-version"]
                  ["vcs" "commit"]
                  ["vcs" "tag" "--no-sign"]
                  ["deploy"]
                  ["vcs" "push"]])
